src/engine/Union.cpp

```diff
@@ -233,4 +237,63 @@ void Union::computeUnion(
     writeColumn(left, targetColumn, leftCol, 0u);
     writeColumn(right, targetColumn, rightCol, left.size());
   }
+  return res;
+}
+
+// _____________________________________________________________________________
+template <bool left>
+std::vector<ColumnIndex> Union::computePermutation() const {
+  ColumnIndex startOfUndefColumns = _subtrees[left ? 0 : 1]->getResultWidth();
+  std::vector<ColumnIndex> permutation(_columnOrigins.size());
+  for (size_t targetColIdx = 0; targetColIdx < _columnOrigins.size();
+       ++targetColIdx) {
+    ColumnIndex originIndex = _columnOrigins.at(targetColIdx)[left ? 0 : 1];
+    if (originIndex == NO_COLUMN) {
```

Comment:
Don'r resize, but maximally reserve the `permutation`, then do a range-for over the `_columnOrigins` with `push_back` on the `permutation`.
That makes the code a little more readable.
And you can precompute the `resIdx = left?0:1` then you don't repeat this pattern (less error-prone).